pkg/boot/linux/load_linux_arm64.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	return chosen, nil
+}
+
+func mmap(f *os.File) (data []byte, ummap func() error, err error) {
+	fd := f.Fd()
+	s, err := f.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("stat error: %w", err)
+	}
+	if s.Size() == 0 {
+		return nil, fmt.Errorf("cannot mmap zero-len file")
+	}
+	d, err := unix.Mmap(int(fd), 0, int(s.Size()), syscall.PROT_READ, syscall.MAP_PRIVATE)
+	if err != nil {
+		return nil, fmt.Errorf("mmap failed: %w", err)
+	}
+
+	ummap = func() error {
+		return unix.Munmap(d)
+	}
+
+	return d, ummap, nil
+}
+
+// KexecLoad loads arm64 Image, with the given ramfs and kernel cmdline.
+func KexecLoad(kernel, ramfs *os.File, cmdline string, opts KexecOptions) error {
+	var err error
+	// kmem is a struct holding kexec segments.
+	//
+	// It has routines to work with physical memory
+	// ranges.
+	var kmem *Memory
+	var kernelRange, ramfsRange, dtbRange, trampolineRange Range
+
+	fdt, err := loadFDT(opts.DTB)
+	if err != nil {
+		return fmt.Errorf("loadFDT(%s) = %v", opts.DTB, err)
+	}
+
+	chosen, err := sanitizeFDT(fdt)
+	if err != nil {
+		return fmt.Errorf("sanitizeFDT(%v) = %v", fdt, err)
+	}
+
+	// Prepare segments.
+	kmem = &Memory{}
+	Debug("Try parsing memory map...")
+	if err := kmem.ParseMemoryMapFromFDT(fdt); err != nil {
+		return fmt.Errorf("ParseMemoryMapFromFDT(%v): %v", fdt, err)
+	}
+
+	var kernelBuf []byte
+	if opts.MmapKernel {
+		Debug("Mmapping kernel to virtual buffer...")
+		kernelBuf, ummap, err := mmap(kernel)
+		if err != nil {
+			return fmt.Errorf("mmap kernel: %v", err)
+		}
+		defer func() {
+			if err = ummap(); err != nil {
+				Debug("Ummap kernel failed: %v", err)
+			}
+		}()
+	} else {
+		Debug("Read kernel from file ...")
+		kernelBuf, err = uio.ReadAll(kernel)
+		if err != nil {
+			return fmt.Errorf("read kernel from file: %v", err)
+		}
+	}
+
+	kImage, err := image.ParseFromBytes(kernelBuf)
+	if err != nil {
+		return fmt.Errorf("parse arm64 Image from bytes: %v", err)
+	}
+
+	if kImage.Header.TextOffset > 0 {
+		// TODO(10000TB): load kernel by text_offset.
+		return fmt.Errorf("non zero text offset is not supported yet.")
+	} else {
+		// TextOffset as 0.
+		if kernelRange, err = kmem.AddKexecSegmentExplicit(b, ah.ImageSize); err != nil {
+			return fmt.Errorf("add kernel segment: %v", err)
+		}
+	}
+	Debug("Added %d byte (size %d) kernel at %s", len(kernelBuf), ah.ImageSize, kernelRange)
+
+	var ramfsBuf []byte
+	if ramfs != nil {
+		if opts.MmapRamfs {
+			Debug("Mmap ramfs file to virtual buffer...")
+			ramfsBuf, ummap, err := mmap(ramfs)
+			if err != nil {
+				return fmt.Errorf("mmap ramfs: %v", err)
+			}
+			defer func() {
+				if err = ummap(); err != nil {
+					return Debug("Ummap ramfs failed: %v", err)
+				}
+			}()
+		} else {
+			Debug("Read ramfs from file...")
+			ramfsBuf, err = uio.ReadAll(ramfs)
+			if err != nil {
+				return fmt.Errorf("read ramfs from file: %v", err)
+			}
+		}
+	}
+
+	// NOTE(10000TB): This need be placed after kernel by convention.
```

Comment:
is that a requirement? You know it does not work if you don't do that?

Reply:
Does not appear so as written for AArch64... https://www.kernel.org/doc/html/latest/arm64/booting.html

But I happened to read it from here, https://www.kernel.org/doc/Documentation/arm/Booting,  then thought to that maybe a convention to keep, 

"If an initramfs is in use then, as with the dtb, it must be placed in
a region of memory where the kernel decompressor will not overwrite it
while also with the region which will be covered by the kernel's
low-memory mapping.

A safe location is just above the device tree blob which itself will
be loaded just above the 128MiB boundary from the start of RAM as
recommended above."

device tree load above kernel, ramfs load above dtb, so above kernel. As we load ramfs before dtb, so that we can save its location into dtb as boot params; so I think that implicitly means  ramfs load above kernel (w/ awareness that it need save space for dtb to load).

But this boot doc seems very old, arm32 ? or ? I could not tell, but the register naming R0, ... seems old.